#  Ejercicios prácticos Pandas

## Preparando  datos en un zip para ser leídos en Pandas

In [1]:
!pwd

/home/dsc/Repos/numpy-pandas-notebook


In [6]:
path_to_folder = '/home/dsc/Data/us_dot/otp/'

In [3]:
! ls {path_to_folder}

last_20.txt				On_Time_On_Time_Performance_2015_5.zip
On_Time_On_Time_Performance_2015_1.zip	On_Time_On_Time_Performance_2015_6.zip
On_Time_On_Time_Performance_2015_2.zip	On_Time_On_Time_Performance_2015_7.zip
On_Time_On_Time_Performance_2015_3.zip	On_Time_On_Time_Performance_2015_8.zip
On_Time_On_Time_Performance_2015_4.zip


In [8]:
! unzip -l {path_to_folder}On_Time_On_Time_Performance_2015_1.zip

Archive:  /home/dsc/Data/us_dot/otp/On_Time_On_Time_Performance_2015_1.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
211633432  2015-04-16 20:02   On_Time_On_Time_Performance_2015_1.csv
    12054  2015-04-16 20:02   readme.html
---------                     -------
211645486                     2 files


_Nota:_ _-l_ me dice el contenido del zip sin descrompimirlo

In [10]:
! unzip -o {path_to_folder}On_Time_On_Time_Performance_2015_1.zip readme.html

Archive:  /home/dsc/Data/us_dot/otp/On_Time_On_Time_Performance_2015_1.zip
  inflating: readme.html             


_Nota_: _-o_ me extrae un archivo concreto. 

In [11]:
! ls -lrt

total 1024
-rw-r--r-- 1 dsc dsc  12054 abr 16  2015 readme.html
-rw-r--r-- 1 dsc dsc     24 ene 11 17:09 README.md
-rw-r--r-- 1 dsc dsc  85345 ene 11 21:49 01_Introduction.ipynb
-rw-r--r-- 1 dsc dsc 157891 ene 18 19:56 Clase_3_1_Numpy-Pandas_JLG.ipynb
-rw-r--r-- 1 dsc dsc   7943 ene 18 20:25 tips.csv
-rw-r--r-- 1 dsc dsc  78782 ene 18 21:47 Clase_3_2_GroupBy_Numpy-Pandas_JLG
drwxr-xr-x 3 dsc dsc   4096 ene 19 11:59 Solved
-rw-r--r-- 1 dsc dsc 684839 ene 19 12:10 Clase_4_1_Plotting_JLG.ipynb
-rw-r--r-- 1 dsc dsc   2350 ene 19 12:13 Clase_4_2_Intro_to_pandas_practical_JLG.ipynb


El readme.html ha sido extraido

In [12]:
! head readme.html

<HTML>
<HEAD>
	<TITLE> readme.html </TITLE>
</HEAD>
<BODY>
	<TABLE><TR><TD COLSPAN=2><H4>BACKGROUND</H4></TD></TR>
	<TR><TD COLSPAN=2>The data contained in the compressed file has been extracted from the 
	On-Time Performance data table of the "On-Time" database from the TranStats data library. 
	The time period is indicated in the name of the compressed file; 
	for example, XXX_XXXXX_2001_1 contains data of the first month of the year 2001. </TD></TR>


In [25]:
from IPython.display import IFrame
IFrame('readme.html', width=950, height=400)

Así se lee desde notebooks una página web. También podría poner la del marca. 

In [22]:
from IPython.display import IFrame
IFrame('http://marca.com', width=950, height=400)

In [29]:
! ! unzip -p {path_to_folder}On_Time_On_Time_Performance_2015_1.zip | head -n 2

"Year","Quarter","Month","DayofMonth","DayOfWeek","FlightDate","UniqueCarrier","AirlineID","Carrier","TailNum","FlightNum","OriginAirportID","OriginAirportSeqID","OriginCityMarketID","Origin","OriginCityName","OriginState","OriginStateFips","OriginStateName","OriginWac","DestAirportID","DestAirportSeqID","DestCityMarketID","Dest","DestCityName","DestState","DestStateFips","DestStateName","DestWac","CRSDepTime","DepTime","DepDelay","DepDelayMinutes","DepDel15","DepartureDelayGroups","DepTimeBlk","TaxiOut","WheelsOff","WheelsOn","TaxiIn","CRSArrTime","ArrTime","ArrDelay","ArrDelayMinutes","ArrDel15","ArrivalDelayGroups","ArrTimeBlk","Cancelled","CancellationCode","Diverted","CRSElapsedTime","ActualElapsedTime","AirTime","Flights","Distance","DistanceGroup","CarrierDelay","WeatherDelay","NASDelay","SecurityDelay","LateAircraftDelay","FirstDepTime","TotalAddGTime","LongestAddGTime","DivAirportLandings","DivReachedDest","DivActualElapsedTime","DivArrDelay","DivDistance","Div1Airport","Div1A

_Nota_: el -p es para ver el contenido del fichero sin descomprimi. Aquí no esdtamos seleccionando ningunmo de dentro, aśi que te muestra todos concatenados. Veamos el final de este resultado, que se corresponde con el final del html

In [28]:
! unzip -p {path_to_folder}On_Time_On_Time_Performance_2015_1.zip | tail -n 2

</BODY>
</HTML>


In [30]:
import os 
import zipfile

_Nota_: el paquete _os_ de _Python_ permite realizar operaciones con el Sistema Operativo. El paquete _zipfile_ permite trabajr con zips

In [33]:
filenames = os.listdir(path_to_folder)
filenames

['On_Time_On_Time_Performance_2015_2.zip',
 'On_Time_On_Time_Performance_2015_8.zip',
 'On_Time_On_Time_Performance_2015_4.zip',
 'On_Time_On_Time_Performance_2015_3.zip',
 'On_Time_On_Time_Performance_2015_7.zip',
 'On_Time_On_Time_Performance_2015_1.zip',
 'On_Time_On_Time_Performance_2015_6.zip',
 'On_Time_On_Time_Performance_2015_5.zip',
 'last_20.txt']

In [34]:
zip_file= zipfile.ZipFile(path_to_folder+filenames[0])
zip_file

<zipfile.ZipFile filename='/home/dsc/Data/us_dot/otp/On_Time_On_Time_Performance_2015_2.zip' mode='r'>

In [35]:
csv,readme = zip_file.filelist

In [36]:
csv

<ZipInfo filename='On_Time_On_Time_Performance_2015_2.csv' compress_type=deflate filemode='-rw-r--r--' file_size=192791767 compress_size=21227521>

In [37]:
readme

<ZipInfo filename='readme.html' compress_type=deflate filemode='-rw-r--r--' file_size=12054 compress_size=2208>

_Nota_: como yo ya sé que el fichero contiene dos subficheros, los puedo obtener de esa forma tan bonita. Si no, tendría una lista y la tendría que recorrer:

In [39]:
zip_file.filelist

[<ZipInfo filename='On_Time_On_Time_Performance_2015_2.csv' compress_type=deflate filemode='-rw-r--r--' file_size=192791767 compress_size=21227521>,
 <ZipInfo filename='readme.html' compress_type=deflate filemode='-rw-r--r--' file_size=12054 compress_size=2208>]

In [40]:
csv_file = zip_file.open(csv.filename)

_Nota_: ahora ya he creado un objteo preparado para ser leido con _pandas_

In [41]:
import pandas as pd
df = pd.read_csv(csv_file)

/home/dsc/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (77,84) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Es el mismo comando que usábamos para leer de un fichero por path.

In [42]:
csv_file.close()
zip_file.close()

_Nota_: para liberar el recurso

In [43]:
df.head()

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate UniqueCarrier  \
0  2015        1      2           7          6  2015-02-07            AS   
1  2015        1      2           7          6  2015-02-07            AS   
2  2015        1      2           7          6  2015-02-07            AS   
3  2015        1      2           7          6  2015-02-07            AS   
4  2015        1      2           7          6  2015-02-07            AS   

   AirlineID Carrier TailNum      ...       Div4TailNum  Div5Airport  \
0      19930      AS  N618AS      ...               NaN          NaN   
1      19930      AS  N778AS      ...               NaN          NaN   
2      19930      AS  N586AS      ...               NaN          NaN   
3      19930      AS  N615AS      ...               NaN          NaN   
4      19930      AS  N471AS      ...               NaN          NaN   

   Div5AirportID  Div5AirportSeqID Div5WheelsOn Div5TotalGTime  \
0            NaN               NaN          NaN            NaN   
1            NaN               NaN          NaN            NaN   
2            NaN               NaN          NaN            NaN   
3            NaN               NaN          NaN            NaN   
4            NaN               NaN          NaN            NaN   

  Div5LongestGTime  Div5WheelsOff Div5TailNum  Unnamed: 109  
0              NaN            NaN         NaN           NaN  
1              NaN            NaN         NaN           NaN  
2              NaN            NaN         NaN           NaN  
3              NaN            NaN         NaN           NaN  
4              NaN            NaN         NaN           NaN  

[5 rows x 110 columns]

* ¡Ya lo tenemos!

##  Tratando los datos con Pandas

In [44]:
df.shape

(429191, 110)

Tiene 429191 filas y 110 columnas

In [45]:

df.size

47211010

Tiene 47211010 _celdas_ (filas por columnas, 429191 x 110 )

Imprimo los tipos de cada columna, inferidos

In [48]:
df.dtypes

Year                    int64
Quarter                 int64
Month                   int64
DayofMonth              int64
DayOfWeek               int64
FlightDate             object
UniqueCarrier          object
AirlineID               int64
Carrier                object
TailNum                object
FlightNum               int64
OriginAirportID         int64
OriginAirportSeqID      int64
OriginCityMarketID      int64
Origin                 object
OriginCityName         object
OriginState            object
OriginStateFips         int64
OriginStateName        object
OriginWac               int64
DestAirportID           int64
DestAirportSeqID        int64
DestCityMarketID        int64
Dest                   object
DestCityName           object
DestState              object
DestStateFips           int64
DestStateName          object
DestWac                 int64
CRSDepTime              int64
                       ...   
Div2WheelsOn          float64
Div2TotalGTime        float64
Div2Longes

Busco el número de elementos no nulos por cada columna

In [50]:
non_null_counts = df.count()
non_null_counts

Year                  429191
Quarter               429191
Month                 429191
DayofMonth            429191
DayOfWeek             429191
FlightDate            429191
UniqueCarrier         429191
AirlineID             429191
Carrier               429191
TailNum               425597
FlightNum             429191
OriginAirportID       429191
OriginAirportSeqID    429191
OriginCityMarketID    429191
Origin                429191
OriginCityName        429191
OriginState           429191
OriginStateFips       429191
OriginStateName       429191
OriginWac             429191
DestAirportID         429191
DestAirportSeqID      429191
DestCityMarketID      429191
Dest                  429191
DestCityName          429191
DestState             429191
DestStateFips         429191
DestStateName         429191
DestWac               429191
CRSDepTime            429191
                       ...  
Div2WheelsOn              14
Div2TotalGTime            14
Div2LongestGTime          14
Div2WheelsOff 

In [51]:
non_null_counts[non_null_counts < 1000]

DivActualElapsedTime    706
DivArrDelay             706
Div1WheelsOff           719
Div1TailNum             719
Div2Airport              14
Div2AirportID            14
Div2AirportSeqID         14
Div2WheelsOn             14
Div2TotalGTime           14
Div2LongestGTime         14
Div2WheelsOff             1
Div2TailNum               1
Div3Airport               0
Div3AirportID             0
Div3AirportSeqID          0
Div3WheelsOn              0
Div3TotalGTime            0
Div3LongestGTime          0
Div3WheelsOff             0
Div3TailNum               0
Div4Airport               0
Div4AirportID             0
Div4AirportSeqID          0
Div4WheelsOn              0
Div4TotalGTime            0
Div4LongestGTime          0
Div4WheelsOff             0
Div4TailNum               0
Div5Airport               0
Div5AirportID             0
Div5AirportSeqID          0
Div5WheelsOn              0
Div5TotalGTime            0
Div5LongestGTime          0
Div5WheelsOff             0
Div5TailNum         

_Ejercicio_:quitar todas las columas con menos de 1000 filas rellenos

In [55]:
df.columns[non_null_counts < 1000]

Index(['DivActualElapsedTime', 'DivArrDelay', 'Div1WheelsOff', 'Div1TailNum',
       'Div2Airport', 'Div2AirportID', 'Div2AirportSeqID', 'Div2WheelsOn',
       'Div2TotalGTime', 'Div2LongestGTime', 'Div2WheelsOff', 'Div2TailNum',
       'Div3Airport', 'Div3AirportID', 'Div3AirportSeqID', 'Div3WheelsOn',
       'Div3TotalGTime', 'Div3LongestGTime', 'Div3WheelsOff', 'Div3TailNum',
       'Div4Airport', 'Div4AirportID', 'Div4AirportSeqID', 'Div4WheelsOn',
       'Div4TotalGTime', 'Div4LongestGTime', 'Div4WheelsOff', 'Div4TailNum',
       'Div5Airport', 'Div5AirportID', 'Div5AirportSeqID', 'Div5WheelsOn',
       'Div5TotalGTime', 'Div5LongestGTime', 'Div5WheelsOff', 'Div5TailNum',
       'Unnamed: 109'],
      dtype='object')

_Nota_: aquí obtengo la lista de todas las columnas con menos de 1000 datos rellenos

In [60]:
df2 = df.drop(df.columns[non_null_counts < 1000],axis=1)

In [61]:
df2.shape

(429191, 73)

Nos hemos quedado sólo con 73 columnas

_Ejercicio_: hacer lo mismo usando dropna

In [64]:
df.dropna(thresh=1000, axis = 1).shape

(429191, 73)

_Ejercicio_: sacar las columnas que contienen origin en el nombre

In [70]:
"hola".upper()

'HOLA'

In [90]:
df.columns.str.lower().str.contains('origin')

array([False, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False])

_Nota_: ojo, que tengo que poner ese doble str

In [93]:
df[df.columns[df.columns.str.lower().str.contains('origin')]].sample(5)

OriginAirportID  OriginAirportSeqID  OriginCityMarketID Origin  \
379314            11618             1161802               31703    EWR   
110211            14771             1477101               32457    SFO   
27242             12478             1247802               31703    JFK   
192667            10397             1039705               30397    ATL   
360009            13930             1393003               30977    ORD   

           OriginCityName OriginState  OriginStateFips OriginStateName  \
379314         Newark, NJ          NJ               34      New Jersey   
110211  San Francisco, CA          CA                6      California   
27242        New York, NY          NY               36        New York   
192667        Atlanta, GA          GA               13         Georgia   
360009        Chicago, IL          IL               17        Illinois   

        OriginWac  
379314         21  
110211         91  
27242          22  
192667         34  
360009         41

Ahora llegamos a la conclusión, por lo que sea, que loas columnas que me interesan son las siguientes:

In [97]:
important_columns = ['FlightDate', 'DayOfWeek', 'Carrier', 'TailNum', 'FlightNum','Origin',
                     'OriginCityName', 'OriginStateName', 'Dest', 'DestCityName', 'DestStateName',
                    'DepTime', 'DepDelay', 'AirTime', 'Distance']

DateOfWeek: De 1 a 7, siendo 1 el dominbgo
Carrier: aerolinea
TailNum:matrícula del avión
AirTime: tiempo que está en el aire. 


In [99]:
df2 = df[important_columns]
df2.sample(5)

FlightDate  DayOfWeek Carrier TailNum  FlightNum Origin  \
117018  2015-02-10          2      AA  N011AA       1079    DFW   
8910    2015-02-23          1      DL  N920DE       2251    CLT   
43696   2015-02-14          6      EV  N840AS       4931    ATL   
369339  2015-02-17          2      OO  N494CA       7391    SLC   
313483  2015-02-17          2      WN  N289CT        536    LAS   

               OriginCityName OriginStateName Dest     DestCityName  \
117018  Dallas/Fort Worth, TX           Texas  ELP      El Paso, TX   
8910            Charlotte, NC  North Carolina  ATL      Atlanta, GA   
43696             Atlanta, GA         Georgia  SGF  Springfield, MO   
369339     Salt Lake City, UT            Utah  GCC     Gillette, WY   
313483          Las Vegas, NV          Nevada  BNA    Nashville, TN   

       DestStateName  DepTime  DepDelay  AirTime  Distance  
117018         Texas   1144.0      -6.0     81.0     551.0  
8910         Georgia   1153.0      -7.0     42.0     226.0  
43696       Missouri   1141.0       6.0     96.0     563.0  
369339       Wyoming   1502.0      -2.0     71.0     410.0  
313483     Tennessee      NaN       NaN      NaN    1587.0

Ahora queremos traducir el DepTime, pueso como número, me lo traduzca en un string tipo HH:MM
Ejemplo: el número 1810.0 debre ser las 18:10

In [108]:
def deptime_to_string(num):
    if num!=num: #NaN
        num = 0.0 
    hour = int(num/100) % 24
    minute = int(num%100) % 24
    return '%02d:%02d' % (hour,minute)

deptime_to_string(810.0)

'08:10'

También queremos arreglar las fechas, que están como strings

In [112]:
ts = pd.to_datetime('2015-01-15 08:30')
ts

Timestamp('2015-01-15 08:30:00')

In [115]:
df2['DepTime'].apply(deptime_to_string).head(5)

0    09:11
1    17:17
2    11:08
3    22:01
4    19:17
Name: DepTime, dtype: object

In [116]:
df2['DepTimeStr'] = df2['DepTime'].apply(deptime_to_string)

/home/dsc/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


_Nota_: te avisa de que hay menos datos, pero no pasa nada, no es juna array, es una serie con sus índices, por lo que lo insertar bien

Queremos crear la nueva columna de tipo datetime. 
Pruebo:

In [120]:
pd.to_datetime(df2['FlightDate']+' ' + df2['DepTimeStr']).head()

0   2015-02-07 09:11:00
1   2015-02-07 17:17:00
2   2015-02-07 11:08:00
3   2015-02-07 22:01:00
4   2015-02-07 19:17:00
dtype: datetime64[ns]

In [121]:
departure_timestamps = pd.to_datetime(df2['FlightDate']+' ' + df2['DepTimeStr'])
df2['DepDateTime'] = departure_timestamps


/home/dsc/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [122]:
df3 = df2.drop(['FlightDate', 'DepTime', 'DepTimeStr'], axis=1)

In [123]:
df3.head()

DayOfWeek Carrier TailNum  FlightNum Origin   OriginCityName  \
0          6      AS  N618AS        131    PDX     Portland, OR   
1          6      AS  N778AS        132    FAI    Fairbanks, AK   
2          6      AS  N586AS        140    ANC    Anchorage, AK   
3          6      AS  N615AS        143    PDX     Portland, OR   
4          6      AS  N471AS        149    LAX  Los Angeles, CA   

  OriginStateName Dest   DestCityName DestStateName  DepDelay  AirTime  \
0          Oregon  ANC  Anchorage, AK        Alaska      19.0    195.0   
1          Alaska  SEA    Seattle, WA    Washington     -13.0    197.0   
2          Alaska  PDX   Portland, OR        Oregon      -8.0    192.0   
3          Oregon  ANC  Anchorage, AK        Alaska      -9.0    193.0   
4      California  ANC  Anchorage, AK        Alaska       1.0    289.0   

   Distance         DepDateTime  
0    1542.0 2015-02-07 09:11:00  
1    1533.0 2015-02-07 17:17:00  
2    1542.0 2015-02-07 11:08:00  
3    1542.0 2015-02-07 22:01:00  
4    2345.0 2015-02-07 19:17:00

###  Hot to rename a column

In [124]:
df.rename(columns={'Carrier': 'CarrierR'})

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate UniqueCarrier  \
0       2015        1      2           7          6  2015-02-07            AS   
1       2015        1      2           7          6  2015-02-07            AS   
2       2015        1      2           7          6  2015-02-07            AS   
3       2015        1      2           7          6  2015-02-07            AS   
4       2015        1      2           7          6  2015-02-07            AS   
5       2015        1      2           7          6  2015-02-07            AS   
6       2015        1      2           7          6  2015-02-07            AS   
7       2015        1      2           7          6  2015-02-07            AS   
8       2015        1      2           7          6  2015-02-07            AS   
9       2015        1      2           7          6  2015-02-07            AS   
10      2015        1      2           7          6  2015-02-07            AS   
11      2015        1      2           7          6  2015-02-07            AS   
12      2015        1      2           7          6  2015-02-07            AS   
13      2015        1      2           7          6  2015-02-07            AS   
14      2015        1      2           7          6  2015-02-07            AS   
15      2015        1      2           7          6  2015-02-07            AS   
16      2015        1      2           7          6  2015-02-07            AS   
17      2015        1      2           7          6  2015-02-07            AS   
18      2015        1      2           7          6  2015-02-07            AS   
19      2015        1      2           7          6  2015-02-07            AS   
20      2015        1      2           7          6  2015-02-07            AS   
21      2015        1      2           7          6  2015-02-07            AS   
22      2015        1      2           7          6  2015-02-07            AS   
23      2015        1      2           7          6  2015-02-07            AS   
24      2015        1      2           7          6  2015-02-07            AS   
25      2015        1      2           7          6  2015-02-07            AS   
26      2015        1      2           7          6  2015-02-07            AS   
27      2015        1      2           7          6  2015-02-07            AS   
28      2015        1      2           7          6  2015-02-07            AS   
29      2015        1      2           7          6  2015-02-07            AS   
...      ...      ...    ...         ...        ...         ...           ...   
429161  2015        1      2          18          3  2015-02-18            OO   
429162  2015        1      2          21          6  2015-02-21            OO   
429163  2015        1      2           2          1  2015-02-02            OO   
429164  2015        1      2           4          3  2015-02-04            OO   
429165  2015        1      2          21          6  2015-02-21            OO   
429166  2015        1      2           8          7  2015-02-08            OO   
429167  2015        1      2           9          1  2015-02-09            OO   
429168  2015        1      2          10          2  2015-02-10            OO   
429169  2015        1      2           3          2  2015-02-03            OO   
429170  2015        1      2           1          7  2015-02-01            OO   
429171  2015        1      2           5          4  2015-02-05            OO   
429172  2015        1      2          11          3  2015-02-11            OO   
429173  2015        1      2           7          6  2015-02-07            OO   
429174  2015        1      2           6          5  2015-02-06            OO   
429175  2015        1      2          14          6  2015-02-14            OO   
429176  2015        1      2          28          6  2015-02-28            OO   
429177  2015        1      2          12          4  2015-02-12            OO   
429178  2015        1      2          1

Ya tenemos todo preparadao para trabajar

#### _Ejercicios_

Calcular los tops 5 retrasos

In [128]:
df3.head()

DayOfWeek Carrier TailNum  FlightNum Origin   OriginCityName  \
0          6      AS  N618AS        131    PDX     Portland, OR   
1          6      AS  N778AS        132    FAI    Fairbanks, AK   
2          6      AS  N586AS        140    ANC    Anchorage, AK   
3          6      AS  N615AS        143    PDX     Portland, OR   
4          6      AS  N471AS        149    LAX  Los Angeles, CA   

  OriginStateName Dest   DestCityName DestStateName  DepDelay  AirTime  \
0          Oregon  ANC  Anchorage, AK        Alaska      19.0    195.0   
1          Alaska  SEA    Seattle, WA    Washington     -13.0    197.0   
2          Alaska  PDX   Portland, OR        Oregon      -8.0    192.0   
3          Oregon  ANC  Anchorage, AK        Alaska      -9.0    193.0   
4      California  ANC  Anchorage, AK        Alaska       1.0    289.0   

   Distance         DepDateTime  
0    1542.0 2015-02-07 09:11:00  
1    1533.0 2015-02-07 17:17:00  
2    1542.0 2015-02-07 11:08:00  
3    1542.0 2015-02-07 22:01:00  
4    2345.0 2015-02-07 19:17:00

In [138]:
df3.sort_values('DepDelay',ascending=False).head(5)

DayOfWeek Carrier TailNum  FlightNum Origin OriginCityName  \
126203          6      AA  N028AA       1312    STL  St. Louis, MO   
117050          7      AA  N007AA       1080    EGE      Eagle, CO   
73864           7      HA  N374HA         50    HNL   Honolulu, HI   
73885           1      HA  N374HA         51    JFK   New York, NY   
131625          4      AA  N3ASAA       1564    PHX    Phoenix, AZ   

       OriginStateName Dest  DestCityName DestStateName  DepDelay  AirTime  \
126203        Missouri  MIA     Miami, FL       Florida    1587.0    131.0   
117050        Colorado  ORD   Chicago, IL      Illinois    1457.0    123.0   
73864           Hawaii  JFK  New York, NY      New York    1433.0    498.0   
73885         New York  HNL  Honolulu, HI        Hawaii    1429.0    684.0   
131625         Arizona  ORD   Chicago, IL      Illinois    1367.0    167.0   

        Distance         DepDateTime  
126203    1068.0 2015-02-28 08:23:00  
117050    1007.0 2015-02-22 14:08:00  
73864     4983.0 2015-02-08 16:23:00  
73885     4983.0 2015-02-09 09:00:00  
131625    1440.0 2015-02-19 05:06:00

In [130]:
df3.corr()

DayOfWeek  FlightNum  DepDelay   AirTime  Distance
DayOfWeek   1.000000   0.015379 -0.018116  0.015865  0.021141
FlightNum   0.015379   1.000000  0.002797 -0.333575 -0.349537
DepDelay   -0.018116   0.002797  1.000000  0.021388  0.020472
AirTime     0.015865  -0.333575  0.021388  1.000000  0.977982
Distance    0.021141  -0.349537  0.020472  0.977982  1.000000

Muy útil: para ver la correlacíon entre cada variable. 

Origen con mayor media de retraso

In [151]:
top = df3.groupby('Origin')["DepDelay"].mean().sort_values(ascending=False).head(5)
top

Origin
TTN    49.747423
ILG    40.789474
UST    39.000000
SMX    32.327273
JLN    31.120000
Name: DepDelay, dtype: float64

In [150]:
top.index[0]

'TTN'

Ahora queremos esa misma tabla pero obtenindo también el count

In [156]:
df3.groupby('Origin')["DepDelay"].agg(['mean','count']).sort_values('mean',ascending=False).head(10)

mean  count
Origin                  
TTN     49.747423    194
ILG     40.789474     19
UST     39.000000     11
SMX     32.327273     55
JLN     31.120000     50
PLN     30.500000     46
MKG     27.500000     46
SCE     26.888889     54
SUX     26.772727     44
OTH     26.071429     14

Ahora quiere que solo nos quedemos con los que tengan más de treinta vuelos

In [160]:
mean_and_counts = df3.groupby('Origin')["DepDelay"].agg(['mean','count'])

only_more_than_30 = mean_and_counts [ mean_and_counts['count'] > 30]
only_more_than_30.sort_values('count').head()

mean  count
Origin                  
MMH      6.583333     36
BGM     17.361111     36
DVL      4.083333     36
APN     14.093023     43
SUX     26.772727     44

In [165]:
highly_delayed = only_more_than_30.sort_values('mean',ascending=False).head(10)
highly_delayed

NameError: name 'hightly_delayed' is not defined

De la lista total de aerolines, sacars las que están en el top cinco de : no sé muy bien que 

In [166]:
df3.join(highly_delayed.head(5), on='Origin', how = 'inner')['Carrier'].unique()

array(['F9', 'MQ', 'OO'], dtype=object)

### Otro

Mostrar ciudades ordenadas por numero de aeropuertos

Tenemos: OriginCityName (Ciudad), Origin(Aeropuerto)